File: algorithms/code/leetcode/lc118_pascals_triangle/lc118_pascals_triangle.py
# https://leetcode.com/problems/pascals-triangle/

import unittest


class Solution:
    def generate(self, numRows: int) -> list[list[int]]:
        if numRows < 0:
            return []
        triangle = []
        for i in range(numRows):
            row: list[int] = []
            triangle.append(row)
            row.append(1)
            if i > 0:
                above = triangle[i - 1]
                for j in range(i - 1):
                    row.append(above[j] + above[j + 1])
                row.append(1)
        return triangle


class TestCode(unittest.TestCase):
    def test_5(self) -> None:
        expected = [[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1]]
        result = Solution().generate(5)
        assert len(expected) == len(result)
        for expected_i, result_i in zip(expected, result, strict=True):
            assert expected_i == result_i

    def test_nothing(self) -> None:
        assert not Solution().generate(-1)

File: algorithms/code/leetcode/lc118_pascals_triangle/lc118_pascals_triangle.rs
// https://leetcode.com/problems/pascals-triangle/
// #easy

pub struct Solution;

impl Solution {
    pub fn generate(num_rows: i32) -> Vec<Vec<i32>> {
        if num_rows < 0 {
            return Vec::new();
        }
        let mut triangle = Vec::new();
        for i in 0..(num_rows as usize) {
            let mut row = Vec::new();
            row.push(1);
            if 0 < i {
                let above: &Vec<i32> = &triangle[i - 1];
                for j in 0..(i - 1) {
                    row.push(above[j] + above[j + 1]);
                }
                row.push(1);
            }
            triangle.push(row);
        }
        triangle
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_5() {
        let expected = vec![
            vec![1],
            vec![1, 1],
            vec![1, 2, 1],
            vec![1, 3, 3, 1],
            vec![1, 4, 6, 4, 1],
        ];
        let result = Solution::generate(5);
        assert_eq!(result, expected);
    }

    #[test]
    fn test_nothing() {
        assert!(Solution::generate(-1).is_empty());
    }
}
